[java] ImplicitSwitchFallThrough NPE in PMD 7.0.0-rc1

The exceptions from my initial post #4499 are corrected with PMD 7 and the snapshot maven plugin.

However, I get a different set of exceptions with PMD 7.0.0:

```
[WARNING] Exception applying rule ImplicitSwitchFallThrough on file file:///home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java, continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:152:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
    at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:232)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
Caused by: java.lang.NullPointerException
    at java.util.Objects.requireNonNull (Objects.java:208)
    at net.sourceforge.pmd.lang.java.rule.internal.DataflowPass$DataflowResult.switchBranchFallsThrough (DataflowPass.java:263)
    at net.sourceforge.pmd.lang.java.rule.errorprone.ImplicitSwitchFallThroughRule.visit (ImplicitSwitchFallThroughRule.java:39)
    at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.acceptVisitor (ASTSwitchStatement.java:20)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:24)
    at net.sourceforge.pmd.lang.rule.AbstractDelegateRule.apply (AbstractDelegateRule.java:238)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:75)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
[WARNING] Exception occurred on node !debug only! [SwitchStatement:152:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)
[WARNING] Exception applying rule ImplicitSwitchFallThrough on file file:///home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java, continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:275:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
    at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:232)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
Caused by: java.lang.NullPointerException
    at java.util.Objects.requireNonNull (Objects.java:208)
    at net.sourceforge.pmd.lang.java.rule.internal.DataflowPass$DataflowResult.switchBranchFallsThrough (DataflowPass.java:263)
    at net.sourceforge.pmd.lang.java.rule.errorprone.ImplicitSwitchFallThroughRule.visit (ImplicitSwitchFallThroughRule.java:39)
    at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.acceptVisitor (ASTSwitchStatement.java:20)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:24)
    at net.sourceforge.pmd.lang.rule.AbstractDelegateRule.apply (AbstractDelegateRule.java:238)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:75)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
[WARNING] Exception occurred on node !debug only! [SwitchStatement:275:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)
[WARNING] Exception applying rule ImplicitSwitchFallThrough on file file:///home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java, continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:97:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
    at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:232)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
Caused by: java.lang.NullPointerException
    at java.util.Objects.requireNonNull (Objects.java:208)
    at net.sourceforge.pmd.lang.java.rule.internal.DataflowPass$DataflowResult.switchBranchFallsThrough (DataflowPass.java:263)
    at net.sourceforge.pmd.lang.java.rule.errorprone.ImplicitSwitchFallThroughRule.visit (ImplicitSwitchFallThroughRule.java:39)
    at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.acceptVisitor (ASTSwitchStatement.java:20)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:24)
    at net.sourceforge.pmd.lang.rule.AbstractDelegateRule.apply (AbstractDelegateRule.java:238)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:75)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
[WARNING] Exception occurred on node !debug only! [SwitchStatement:97:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)
[WARNING] Exception applying rule ImplicitSwitchFallThrough on file file:///home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java, continuing with next rule
org.apache.commons.lang3.exception.ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:330:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
    at net.sourceforge.pmd.util.AssertionUtil.contexted (AssertionUtil.java:232)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:77)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
Caused by: java.lang.NullPointerException
    at java.util.Objects.requireNonNull (Objects.java:208)
    at net.sourceforge.pmd.lang.java.rule.internal.DataflowPass$DataflowResult.switchBranchFallsThrough (DataflowPass.java:263)
    at net.sourceforge.pmd.lang.java.rule.errorprone.ImplicitSwitchFallThroughRule.visit (ImplicitSwitchFallThroughRule.java:39)
    at net.sourceforge.pmd.lang.java.ast.ASTSwitchStatement.acceptVisitor (ASTSwitchStatement.java:20)
    at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.acceptVisitor (AbstractJavaNode.java:38)
    at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply (AbstractJavaRule.java:24)
    at net.sourceforge.pmd.lang.rule.AbstractDelegateRule.apply (AbstractDelegateRule.java:238)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.applyOnIndex (RuleApplicator.java:75)
    at net.sourceforge.pmd.lang.rule.internal.RuleApplicator.apply (RuleApplicator.java:55)
    at net.sourceforge.pmd.RuleSets.apply (RuleSets.java:158)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.processSource (PmdRunnable.java:140)
    at net.sourceforge.pmd.lang.impl.PmdRunnable.run (PmdRunnable.java:80)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:539)
    at java.util.concurrent.FutureTask.run (FutureTask.java:264)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:833)
[WARNING] Exception occurred on node !debug only! [SwitchStatement:330:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)
[WARNING] There are 4 PMD processing errors:
[WARNING] /home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java: ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:152:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
/home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java: ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:275:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
/home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java: ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:97:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
/home/luc/sources/eclipse/hipparchus/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java: ContextedRuntimeException: java.lang.NullPointerException
Exception Context:
	[1:Rule applied on node=!debug only! [SwitchStatement:330:17]switch (previous) {
                    case UNINITIALIZED :
                        // we are initializing the first point
               (truncated)]
---------------------------------
[INFO] PMD version: 7.0.0-SNAPSHOT
```

Note that I did not fix the warnings from my 6.x.x version of `mpd-ruleset.xml` before running the check, but I doubt these warning are related to the exceptions.

_Originally posted by @maisonobe in https://github.com/pmd/pmd/issues/4499#issuecomment-1519563563_

See source file: https://github.com/Hipparchus-Math/hipparchus/blob/update-build-dependencies/hipparchus-ode/src/main/java/org/hipparchus/ode/events/FilterType.java

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -421,6 +421,7 @@ Language specific fixes:
     * [#4356](https://github.com/pmd/pmd/pull/4356):   \[java] Fix NPE in CloseResourceRule
     * [#4449](https://github.com/pmd/pmd/issues/4449): \[java] AvoidAccessibilityAlteration: Possible false positive in AvoidAccessibilityAlteration rule when using Lambda expression
     * [#4493](https://github.com/pmd/pmd/issues/4493): \[java] MissingStaticMethodInNonInstantiatableClass: false-positive about @<!-- -->Inject
+    * [#4505](https://github.com/pmd/pmd/issues/4505): \[java] ImplicitSwitchFallThrough NPE in PMD 7.0.0-rc1
 * java-multithreading
     * [#2537](https://github.com/pmd/pmd/issues/2537): \[java] DontCallThreadRun can't detect the case that call run() in `this.run()`
     * [#2538](https://github.com/pmd/pmd/issues/2538): \[java] DontCallThreadRun can't detect the case that call run() in `foo.bar.run()`

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/DataflowPass.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/DataflowPass.java
@@ -36,6 +36,7 @@ import net.sourceforge.pmd.lang.java.ast.ASTConstructorCall;
 import net.sourceforge.pmd.lang.java.ast.ASTConstructorDeclaration;
 import net.sourceforge.pmd.lang.java.ast.ASTContinueStatement;
 import net.sourceforge.pmd.lang.java.ast.ASTDoStatement;
+import net.sourceforge.pmd.lang.java.ast.ASTEnumConstant;
 import net.sourceforge.pmd.lang.java.ast.ASTExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTFieldAccess;
 import net.sourceforge.pmd.lang.java.ast.ASTFieldDeclaration;
@@ -954,6 +955,7 @@ public final class DataflowPass {
 
         @Override
         public SpanInfo visitTypeDecl(ASTAnyTypeDeclaration node, SpanInfo data) {
+            // process initializers and ctors first
             processInitializers(node.getDeclarations(), data, node.getSymbol());
 
             for (ASTBodyDeclaration decl : node.getDeclarations()) {
@@ -989,12 +991,14 @@ public final class DataflowPass {
 
             for (ASTBodyDeclaration declaration : declarations) {
                 final boolean isStatic;
-                if (declaration instanceof ASTFieldDeclaration) {
+                if (declaration instanceof ASTEnumConstant) {
+                    isStatic = true;
+                } else if (declaration instanceof ASTFieldDeclaration) {
                     isStatic = ((ASTFieldDeclaration) declaration).isStatic();
                 } else if (declaration instanceof ASTInitializer) {
                     isStatic = ((ASTInitializer) declaration).isStatic();
                 } else if (declaration instanceof ASTConstructorDeclaration
-                        || declaration instanceof ASTCompactConstructorDeclaration) {
+                    || declaration instanceof ASTCompactConstructorDeclaration) {
                     ctors.add(declaration);
                     continue;
                 } else {

--- a/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ImplicitSwitchFallThrough.xml
+++ b/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/ImplicitSwitchFallThrough.xml
@@ -400,4 +400,25 @@ record MyRecord(boolean b) {
 }
         ]]></code>
     </test-code>
+
+    <test-code>
+        <description>NPE with Switch in enum constant body</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>7</expected-linenumbers>
+        <code><![CDATA[
+            enum Foo {
+                A {
+                    void foo() {
+                        switch (A) {
+                        case A:
+                            if (A != null) {}
+                        default:
+                            throw new UnsupportedOperationException();
+                        }
+                    }
+                }
+            }
+
+            ]]></code>
+    </test-code>
 </test-data>

